mixer/adapter/svcctrl/quotaprocessor_test.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package svcctrl
+
+import (
+	"context"
+	"reflect"
+	"testing"
+	"time"
+
+	pbtypes "github.com/gogo/protobuf/types"
+	rpc "github.com/googleapis/googleapis/google/rpc"
+	sc "google.golang.org/api/servicecontrol/v1"
+
+	"istio.io/istio/mixer/adapter/svcctrl/config"
+	"istio.io/istio/mixer/pkg/adapter"
+	at "istio.io/istio/mixer/pkg/adapter/test"
+	"istio.io/istio/mixer/pkg/status"
+	"istio.io/istio/mixer/template/quota"
+)
+
+type quotaTest struct {
+	processor *quotaImpl
+	client    *mockSvcctrlClient
+	instance  quota.Instance
+}
+
+type quotaTestCase struct {
+	args            adapter.QuotaArgs
+	expectedRequest string
+	response        *sc.AllocateQuotaResponse
+	expectedResult  adapter.QuotaResult
+}
+
+func setupQuotaTest(t *testing.T) *quotaTest {
+	svcCfg := &config.GcpServiceSetting{
+		Quotas: []*config.Quota{
+			{
+				"ratelimit.quota.istio-system",
+				"read-requests",
+				&pbtypes.Duration{
+					Seconds: 60,
+				},
+			},
+		},
+	}
+
+	mockClient := &mockSvcctrlClient{}
+	ctx := &handlerContext{
+		env: at.NewEnv(t),
+		serviceConfigIndex: map[string]*config.GcpServiceSetting{
+			svcCfg.MeshServiceName: svcCfg,
+		},
+		client: mockClient,
+	}
+
+	proc, err := newQuotaProcessor(svcCfg.MeshServiceName, ctx)
+	if err != nil {
+		t.Fatalf("fail to create test quota processor: %v", err)
+	}
+
+	return &quotaTest{
+		proc,
+		mockClient,
+		quota.Instance{
+			Name: "ratelimit.quota.istio-system",
+			Dimensions: map[string]interface{}{
+				"api_key":       "api-key",
+				"api_operation": "echo",
+			},
+		},
+	}
+}
+
+func TestProcessQuota(t *testing.T) {
+	testCases := []*quotaTestCase{
+		{
+			adapter.QuotaArgs{
+				"-1-",
```

Comment:
I found this "-1-" is really difficult to understand. Poor design.

Not your issue, though.